### PR TITLE
Add direct URL conversion to lockfile

### PR DESCRIPTION
**Pull request by charliermarsh**
## Summary

Similar to #3630, but for direct URL distributions (for both wheels and source distributions).

### CODE REVIEW

**Comment by charliermarsh** [comment]
We might want some dedicated `LockUrl` types here to facilitate all these conversions. It's a lot of conversion code and the expectations around the URLs aren't always clear (i.e., is it the plan "location" URL? or does it include adornments, like the subdirectory?).

File: crates/uv-resolver/src/lock.rs [hunk 1]
```diff
@@ -256,8 +256,31 @@ impl Distribution {
...
-                // TODO: Handle other kinds of sources.
-                _ => todo!(),
+                SourceKind::Direct(direct) => {
+                    let filename: WheelFilename = self.wheels[best_wheel_index].filename.clone();
+
+                    // Reconstruct the URL with the subdirectory query parameter.
+                    let mut url = self.id.source.url.clone();
+                    if let Some(subdirectory) = direct.subdirectory.as_ref() {
+                        url.query_pairs_mut()
+                            .append_pair("subdirectory", subdirectory.as_str());
```
Comment (konstin) [inline crates/uv-resolver/src/lock.rs:266]: Shouldn't this be fragment not query?

Comment (charliermarsh) [reply crates/uv-resolver/src/lock.rs:266]: Not necessarily. This format is internal to the lockfile. We’re using the fragment for the Git SHA.

**Review by BurntSushi — APPROVED** [review]
LGTM! Thank you!

File: crates/uv-resolver/src/lock.rs (hunk 1 above)
Comment (konstin) [reply crates/uv-resolver/src/lock.rs:266]: Isn't this method the reverse conversion, from lockfile to installable `Dist`?

Comment (charliermarsh) [reply crates/uv-resolver/src/lock.rs:266]: Yes you're right.

**Pull request merged**